### PR TITLE
JK Memory estimates and tighter bounds on DFT cache

### LINKED ISSUE #1218: doc: The manual is hiding the good stuff from us.
As discussed on slack today (5th Sept 2018, 12:00 UTC), there are a few points where the docs are inaccurate:

1) SOSCF with DFT *is* supported; as @dgasmith said, the feature is not thoroughly checked - eg. DAMP and SOSCF should be exclusive.:
https://github.com/psi4/psi4/blob/master/doc/sphinxman/source/scf.rst#second-order-convergence

2) Gradients with LRC and DF integrals *are* supported, as reported by @hokru and confirmed by @dgasmith:
https://github.com/psi4/psi4/blob/b0b8cd19f5a360871277cbcc00193c88c00ec3e9/doc/sphinxman/source/dft.rst#eri-algorithms

Please add more issues below. I currently (Autumn 2018) don't plan to make a PR to fix these.

### PR DESCRIPTION
**Pull request by dgasmith**
## Description
Several bug fixes and enhancements promised for 1.3. Primary goal for these changes is to localize the estimation of DF memory in a single location to prevent some odd errors.

## Todos
Notable points (developer or user-interest) that this PR has or will accomplish.
- [x] Remove class method attribute for `Wavefunction.to_file`.
- [x] Reworks SCF `save_orbitals` to use canonical Wavefunction serialization.
- [x] `JK.name()` returns the name of the JK object.
- [x] `JK.memory_estimate()` attempts to estimate the amount of in-core memory required.
- [x] Reworks DFT cache size logic to account for Schwarz screening.
- [x] Removes SCF initialize JK from C++ for better initialization control
- [x] Attempts to close #1218.

## Questions
 - [ ] Can we get better estimates on PK memory usage @jgonthier.
 - [ ] Does IWL use virtually zero memory? @andysim 

## Checklist
- [x] Tests added for any new features
- [x] [All or relevant fraction of full tests run](http://psicode.org/psi4manual/master/build_planning.html#how-to-run-a-subset-of-tests)

## Status
- [x] Ready for review
- [x] Ready for merge

### CODE REVIEW

**Comment by JonathonMisiewicz** [comment]
Disclaimer that I have a PR that _also_ rewrites orbital reading/writing to use wavefunction serialization.

My PR will be part of a larger change of the orbital reading tech to account for loss of symmetry (which also involves accounting for axis permutations), so I'm fine with putting the wavefunction serialization change in a non-feature PR. If it matters, my code is passing intense field tests, but I'd rather not make the PR until I have the bugs ironed out. I know for sure that one bug is due to a libmints bug that @jturney is investigating. My guess is that is the last bug, but famous last words.

**Comment by dgasmith** [comment]
@JonathonMisiewicz Thanks for the note, does hit impact your revisions or are the fairly comparable?

@psi4/maintainers This is ready for review.

File: tests/pytest/test_jkmemory.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+"""
+Tests for the memory estimators on JK objects
+"""
+
+import psi4
+import pytest
+from utils import *
+
```
Comment (loriab) [inline tests/pytest/test_jkmemory.py]: add https://github.com/psi4/psi4/blob/master/tests/pytest/test_qcvars.py#L9 so it always runs in CI

File: tests/serial-wfn/input.dat [hunk 2]
```diff
@@ -17,15 +17,15 @@ set {
 e, wfn_old = energy('hf', return_wfn=True)
 
 # write the wavefunction to file
-Wavefunction.to_file(wfn_old, 'my_wfn')
+wfn_old.to_file('my_wfn')
```
Comment (loriab) [inline tests/serial-wfn/input.dat]: more predictable, nice

**Comment by JonathonMisiewicz** [comment]
@dgasmith About as comparable as I could expect.

I am glad to see the `gbs` stripping lines gone. I wasted way too much time trying to figure out if I needed those or not.

File: tests/pytest/test_jkmemory.py (hunk 1 above)
Comment (dgasmith) [reply tests/pytest/test_jkmemory.py]: Not particularly quick, ok to run later?

File: psi4/driver/p4util/python_helpers.py [hunk 3]
```diff
@@ -130,17 +130,32 @@ def _core_wavefunction_build(mol, basis=None):
...
+        If a str reads a Wavefunction from a disk otherwise, assumes the data
+        is passed in.
+
+    Returns
+    -------
+    Wavefunction
+        A deserialized Wavefunction object
+    """
     # load the wavefunction from file
     if isinstance(wfn_data, str):
```
[the remark this answers is not in the document]
Comment (dgasmith) [reply psi4/driver/p4util/python_helpers.py]: Sure.

File: tests/pytest/test_jkmemory.py (hunk 1 above)
Comment (loriab) [reply tests/pytest/test_jkmemory.py]: sure. no label means it runs in "full" (default) and "long" modes.

File: tests/serial-wfn/input.dat (hunk 2 above)
Comment (dgasmith) [reply tests/serial-wfn/input.dat]: BTW @zachglick.

File: psi4/src/psi4/lib3index/fittingmetric.cc [hunk 4]
```diff
@@ -432,103 +432,15 @@ void FittingMetric::form_eig_inverse(double tol) {
...
-                nsig++;
-                eigval[ind] = 1.0 / sqrt(eigval[ind]);
-            }
-            // scale one set of eigenvectors by the diagonal elements j^{-1/2}
-            C_DSCAL(n, eigval[ind], Wp[ind], 1);
-        }
-        delete[] eigval;
-
-        C_DGEMM('T', 'N', n, n, n, 1.0, Jcopyp[0], n, Wp[0], n, 0.0, J[0], n);
-    }
```
Comment (andysim) [inline psi4/src/psi4/lib3index/fittingmetric.cc]: Nice!

File: psi4/src/psi4/libfock/jk.h [hunk 5]
```diff
@@ -377,6 +379,7 @@ class PSI_API JK {
      *        ignored if possible
      */
     void set_cutoff(double cutoff) { cutoff_ = cutoff; }
+    double get_cutoff() { return cutoff_; }
```
Comment (andysim) [inline psi4/src/psi4/libfock/jk.h]: I know we're inconsistent throughout but this function should be marked `const`

File: psi4/src/psi4/libfock/jk.h [hunk 6]
```diff
@@ -393,6 +396,8 @@ class PSI_API JK {
      *        run with their original maximum number)
      */
     void set_omp_nthread(int omp_nthread) { omp_nthread_ = omp_nthread; }
+    int get_omp_nthread() { return omp_nthread_; }
```
Comment (andysim) [inline psi4/src/psi4/libfock/jk.h]: this one should be `const` also

Comment (dgasmith) [reply psi4/src/psi4/libfock/jk.h]: `const` both return and function?

Comment (andysim) [inline psi4/src/psi4/libfock/jk.h]:
```suggestion
    int get_omp_nthread() const { return omp_nthread_; }
```

**Pull request merged**